credentials/tls/certprovider/store_test.go

```diff
@@ -90,262 +100,297 @@ func (c *fakeStableConfig) Canonical() []byte {
+		t.Fatal(err)
+	}
+
+	// Load key material from testdata directory, push it into out fakeProvider
+	// and attempt to read from the Provider returned by the store.
+	testKM1 := loadKeyMaterials(t, "x509/server1_cert.pem", "x509/server1_key.pem", "x509/client_ca_cert.pem")
+	fakeProv.newKeyMaterial(testKM1, nil)
+	if err := readAndVerifyKeyMaterial(prov, testKM1); err != nil {
+		t.Fatal(err)
 	}
 
-	// Close the provider and retry the KeyMaterial() call, and expect it to
-	// fail with a known error.
-	prov.Close()
-	if _, err := prov.KeyMaterial(ctx); err != errProviderClosed {
-		t.Fatalf("provider.KeyMaterial() = %v, wantErr: %v", err, errProviderClosed)
+	// Push new key material and read from the Provider. This should returned
+	// updated key material.
+	testKM2 := loadKeyMaterials(t, "x509/server2_cert.pem", "x509/server2_key.pem", "x509/client_ca_cert.pem")
+	fakeProv.newKeyMaterial(testKM2, nil)
+	if err := readAndVerifyKeyMaterial(prov, testKM2); err != nil {
+		t.Fatal(err)
 	}
 }
 
-// TestStoreWithSingleProviderWithSharing creates multiple instances of the same
-// type of provider through the store (and expects the store's sharing mechanism
-// to kick in) and calls methods on it.
-func (s) TestStoreWithSingleProviderWithSharing(t *testing.T) {
-	prov1, fp1 := makeProvider(t, fakeProvider1Name, fakeConfig)
-	prov2, _ := makeProvider(t, fakeProvider1Name, fakeConfig)
+// TestStoreSingleProviderSameConfigDifferentOpts creates multiple providers of
+// same type, for same configs but different keyMaterial options through the
+// store (and expects the store's sharing mechanism to kick in) and calls
+// methods on them.
+func (s) TestStoreSingleProviderSameConfigDifferentOpts(t *testing.T) {
+	// Create three readers on the same fake provider. Two of these readers use
+	// certName `foo`, while the third one uses certName `bar`.
+	optsFoo := KeyMaterialOptions{CertName: "foo"}
+	optsBar := KeyMaterialOptions{CertName: "bar"}
+	provFoo1, err := GetProvider(fakeProvider1Name, fakeConfig, optsFoo)
+	if err != nil {
+		t.Fatalf("GetProvider(%s, %s, %v) failed: %v", fakeProvider1Name, fakeConfig, optsFoo, err)
+	}
+	defer provFoo1.Close()
+	provFoo2, err := GetProvider(fakeProvider1Name, fakeConfig, optsFoo)
+	if err != nil {
+		t.Fatalf("GetProvider(%s, %s, %v) failed: %v", fakeProvider1Name, fakeConfig, optsFoo, err)
+	}
+	defer provFoo2.Close()
+	provBar1, err := GetProvider(fakeProvider1Name, fakeConfig, optsBar)
+	if err != nil {
+		t.Fatalf("GetProvider(%s, %s, %v) failed: %v", fakeProvider1Name, fakeConfig, optsBar, err)
+	}
+	defer provBar1.Close()
 
-	// Push key materials into the fake provider1.
-	wantKM, err := loadKeyMaterials()
+	// Our fakeProviderBuilder pushes newly created providers on a channel.
+	// Grab the fake provider for optsFoo.
+	p, err := fpb1.providerChan.Receive()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatalf("Timeout when expecting certProvider %q to be created", fakeProvider1Name)
 	}
```

Comment:
Not related to this PR but.. It would be more straightforward and useful in more scenarios if `Receive` (and `Send`) accepted a context, vs. having a hidden internal timeout.  Can this be changed?

Also, we have two copies of this `testutils.Channel`: one in `internal` and one in `internal/xds` -- we need to unify them.

Reply:
Will do in a follow up PR. Definitely makes sense.